### PR TITLE
mgr/dashboard: fix clicked top level menu collapsing issue

### PR DESCRIPTION
**Pull request by PlabanKr**
This fix checks the url after the `NavigationEnd` triggers and changes the `displayedSubMenu` accordingly which opens/collapses desired top level menu.

Fixes: https://tracker.ceph.com/issues/44860
Signed-off-by: Plaban Kumar Mondal <plaban.kr.mondal00@gmail.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Review by ljflores — COMMENTED** [review]
Several linting errors:
```
[lint:tslint  ] Lint errors found in the listed files.
[lint:tslint  ] Linting "ceph-dashboard"...
[lint:tslint  ] /home/jenkins-build/build/workspace/ceph-pull-requests/src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:2:10
[lint:tslint  ] ERROR: 2:10    ordered-imports  Named imports must be alphabetized.
[lint:tslint  ] ERROR: 96:34   quotemark        " should be '
[lint:tslint  ] ERROR: 96:74   quotemark        " should be '
[lint:tslint  ] ERROR: 97:37   quotemark        " should be '
[lint:tslint  ] ERROR: 97:79   quotemark        " should be '
[lint:tslint  ] ERROR: 98:37   quotemark        " should be '
[lint:tslint  ] ERROR: 98:75   quotemark        " should be '
[lint:tslint  ] ERROR: 99:37   quotemark        " should be '
[lint:tslint  ] ERROR: 99:81   quotemark        " should be '
[lint:tslint  ] ERROR: 100:37  quotemark        " should be '
[lint:tslint  ] ERROR: 100:76  quotemark        " should be '
[lint:tslint  ] ERROR: 105:34  quotemark        " should be '
[lint:tslint  ] ERROR: 105:79  quotemark        " should be '
[lint:tslint  ] ERROR: 106:37  quotemark        " should be '
[lint:tslint  ] ERROR: 111:34  quotemark        " should be '
[lint:tslint  ] ERROR: 111:79  quotemark        " should be '
[lint:tslint  ] ERROR: 112:37  quotemark        " should be '
```

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts [hunk 1]
```diff
@@ -78,12 +80,41 @@ export class NavigationComponent implements OnInit, OnDestroy {
...
     this.subs.unsubscribe();
   }
 
+  setDisplayedSubMenu(): void {
+    if (
+      this.router.url.startsWith('/hosts') || this.router.url.startsWith('/inventory')
+      || this.router.url.startsWith('/monitor') || this.router.url.startsWith('/services')
+      || this.router.url.startsWith('/osd') || this.router.url.startsWith('/configuration')
+      || this.router.url.startsWith('/crush-map') || this.router.url.startsWith('/mgr-modules')
+      || this.router.url.startsWith('/logs') || this.router.url.startsWith('/monitoring')
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:100]: There is too much repetition here. Maybe:
```
url = '/monitor/blah'
let pages = ['/hosts','/inventory','/monitor','/services','/osd','/configuration','/crush-map','/mgr-modules','/logs','/monitoring']
let res = pages.some((page) => url.startsWith(page));
`

**Review by pereman2 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:100]: I agree! Also if we change the url to start with `/cluster/hosts`, `/cluster/inventory` etc then we can just eliminate all the if conditions and just check 
```
if this.router.url.startsWith('/cluster/') 
    displaySubmenu = cluster`
```

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts [hunk 2]
```diff
@@ -78,12 +80,41 @@ export class NavigationComponent implements OnInit, OnDestroy {
...
+    ) {
+      this.displayedSubMenu = 'cluster';
+    }
+    if (
+      this.router.url.startsWith('/block/rbd/') || this.router.url.startsWith('/block/mirroring')
+      || this.router.url.startsWith('/block/iscsi')
+    ) {
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:107]:
```suggestion
    if (this.router.url.startsWith('/block/')) {
```

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts [hunk 3]
```diff
@@ -78,12 +80,41 @@ export class NavigationComponent implements OnInit, OnDestroy {
...
+      this.displayedSubMenu = 'block';
+    }
+    if (
+      this.router.url.startsWith('/rgw/daemon') || this.router.url.startsWith('/rgw/user')
+      || this.router.url.startsWith('/rgw/bucket')
+    ) {
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:113]:
```suggestion
    if (this.router.url.startsWith('/rgw/')) {
```

**Review by nizamial09 — COMMENTED** [review]
@PlabanKr Nice work! Some changes from me! Also can you cover this with some unit tests.

**Comment by nizamial09** [comment]
@PlabanKr if you are running [ceph-dev](https://github.com/rhcs-dashboard/ceph-dev), you can set up a pre-commit hook which is described in the readme itself that'll help fix some of the linting issues for you! or you can run `npm run fix` manually.

**Comment by PlabanKr** [comment]
@pereman2  @nizamial09 thank you both for the suggestions. I will try to fix this asap. Regarding the unit test, I am a bit new but I can try to cover this with unit tests.

File: src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts (hunk 2 above)
Comment (PlabanKr) [reply src/pybind/mgr/dashboard/frontend/src/app/core/navigation/navigation/navigation.component.ts:107]: @nizamial09 the routes inside Cluster menu doesn't start with `/cluster/`.
I am thinking of doing this.
```js
let pages = ['/hosts', '/inventory', '/monitor', '/services', '/osd' ... ];
let res = pages.some((page) => this.router.url.startsWith(page));
if (res)  this.displayedSubMenu = 'cluster';
```

**Comment by pereman2** [comment]
Hi @PlabanKr, what is the status of this PR?

**Pull request closed without merging**